### DISCUSSION TITLE
Support for Clearing the last error Trajectory

### DISCUSSION THREAD

**Discussion by nhannt69** [Yaskawa-Global/YMConnect#14] [General]
ChatGPT
Hi everyone,

Thank you for developing this SDK; it has greatly supported our team in working with the YRC 1000 series in our organization.

Currently, we are encountering an issue when working with the trajectory:

When attempting to add an out-of-limit position to the trajectory, the robot shows an alarm "DESTINATION PULSE LIMIT". After resetting the alarm, even when sending the correct position, the previous alarm persists. This only clears after about 5 minutes with an 8015 Timeout [112] alarm.

This error also occurs if we call controller->MotionManager->MotionStop();. I've tried several approaches, such as controller->MotionManager->ClearAllTrajectory();, but the last trajectory remains active.

The only effective solutions I've found are to either reset the robot, reset the TCP or wait in 5 minutes.

I'm asking if there's a better way to clear the cache or reset the last trajectory that was not completed.

Thank you all so much, and have a nice day.

Below:
(1) Setting the motion
(2) Clearing any alarms or stops

(1)
```
if (ui_state.START_Flag)
{
    std::ostringstream stent; stent << "Start Moving: ";
    for (int jn = 0; jn < ui_state.coumove; ++jn) {
        // setting the position
        ui_state.b1crpos->coordinateType = CoordinateType::RobotCoordinate;
        for (int ic = 0; ic < 6; ++ic) {
            ui_state.b1crpos->axisData[ic] = ui_state.rbpos[jn][ic];
        }
        // execute 0: Linear, 1: Circular, 2: Joint, 3: Mid-Cir
        if (ui_state.movTypes[jn] == 0)
        {
            *ui_state.tpstatus = controller->MotionManager->AddPointToTrajectory(LinearMotion(ControlGroupId::R1, *ui_state.b1crpos, ui_state.spdlinear));
            stent << "; Add point " + std::to_string(jn) << " : "; stent << *ui_state.tpstatus;
        }
        else if (ui_state.movTypes[jn] == 2)
        {
            *ui_state.tpstatus = controller->MotionManager->AddPointToTrajectory(JointMotion(ControlGroupId::R1, *ui_state.b1crpos, ui_state.spdjoint));
            stent << "; Add point " + std::to_string(jn) << " : "; stent << *ui_state.tpstatus;
        }
        else if (ui_state.movTypes[jn] == 1) // circular
        {
            CoordinateArray coorarr;
            for (int im = 0; im < 6; ++im) {
                coorarr[im] = ui_state.rbpos[jn - 1][im]; // corrdinate of the previous point
            }
            *ui_state.tpstatus = controller->MotionManager->AddPointToTrajectory(CircularMotion(ControlGroupId::R1, *ui_state.b1crpos, coorarr, ui_state.spdlinear));
            stent << "; Add point " + std::to_string(jn) << " : "; stent << *ui_state.tpstatus;
        }
        else if (ui_state.movTypes[jn] == 3) { continue; } // skip as mid point
    }
    *ui_state.tpstatus = controller->ControlCommands->SetServos(SignalStatus::ON);
    *ui_state.tpstatus = controller->MotionManager->MotionStart();
    stent << *ui_state.tpstatus; stent << "; Start Move";
    *sttlogs << stent.str();
}
```

(2)
```
std::unique_ptr<StatusInfo> temptstt = std::make_unique<StatusInfo>();
*temptstt = controller->Faults->ClearAllFaults();
if (temptstt->StatusCode == 0) { resultmsg.str(" "); }

*ui_state.tpstatus = controller->MotionManager->ClearAllTrajectory();
*ui_state.tpstatus = controller->MotionManager->ClearGroupTrajectory(ControlGroupId::R1);
controller->ControlCommands->SetServos(SignalStatus::OFF);
*ui_state.tpstatus = controller->MotionManager->MotionStart();
*sttlogs << "trying to clear trajectory";
```

**Comment by ted-miller** [comment]
Hi @NTN-hacker. Sorry for the delayed response.

Calling the function `controller->MotionManager->ClearAllTrajectory()` should reset the stored trajectory as you expected.

I don't know if we have tested the behavior after the specific error `DESTINATION PULSE LIMIT`. We'll do some testing this week to try to reproduce this exact error. 

In the meantime, please try the latest [version 1.0.5](https://github.com/Yaskawa-Global/YMConnect/releases/tag/v1.0.5). Changes were made to the MotionManager which might affect this behavior. 

> The only effective solutions I've found are to either reset the robot, reset the TCP or wait in 5 minutes

Can you clarify specifically how you are resetting the robot and/or TCP. I'm not sure what this means.

**Reply by nhannt69** [reply]
Thank you so much. You're so rock!
After applying with version 1.0.5, everything work perfectly without any problems.
ClearAllTrajectory could delete all current trajectory inside the robot.

**Reply by ted-miller** [reply]
Glad to hear the problem is resolved.

If you have any other issues, please open a new discussion.
